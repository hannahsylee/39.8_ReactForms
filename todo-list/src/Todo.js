import React, { useState } from "react";

const Todo = ({ id, task, deleteTodo, editTodo }) => {
  const [editTask, setEditTask] = useState(task);
  const [isEditing, setIsEditing] = useState(false);

  // set isEditing the opposite of what it is before
  const toggleEdit = () => {
    setIsEditing(edit => !edit);
  }

  // set editTodo to the value of what is the event's value?
  const handleChange = evt => {
    setEditTask(evt.target.value);
  }

  const handleDelete = () => deleteTodo(id);

  const handleUpdate = evt => {
    evt.preventDefault();
    editTodo(id, editTask);
    setIsEditing(false);
  };

  // set default settings
  let formSettings = (
    <div>
      <li>{task}</li>
      <button onClick={handleDelete}>X</button>
      <button onClick={toggleEdit}>Edit Task</button>
    </div>
  );

  // set edit settings -> show form for editing when isEditing is True
  if (isEditing){
    formSettings = (
      <div>
        <form onSubmit={handleUpdate}>
          <input
            id="task"
            type="text"
            name="task"
            value={editTask}
            onChange={handleChange}
          />
        <button>Update!</button>
        </form>
      </div>
    )
  };

  return formSettings;

}

export default Todo;


