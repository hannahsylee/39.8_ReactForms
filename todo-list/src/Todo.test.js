import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Todo from "./Todo";

it("renders without crashing", function () {
    render(<Todo />);
});

it("matches snapshot", function () {
    const { asFragment } = render(<Todo />);
    expect(asFragment()).toMatchSnapshot();
});

it("matches snapshot when editing", function() {
    const { asFragment, getByText } = render(<Todo />);
    const editButton = getByText("Edit Task");
    fireEvent.click(editButton);
    expect(asFragment()).toMatchSnapshot();
});

it("runs the update function on form submit", function() {
    const updateMock = jest.fn();
    const { getByText } = render(<Todo editTodo={updateMock} />);
    const editButton = getByText("Edit Task");
    fireEvent.click(editButton);
    const updateButton = getByText("Update!");
    fireEvent.click(updateButton);
    expect(updateMock).toHaveBeenCalled();
});

it("runs the delete function on button click", function() {
    const deleteMock = jest.fn();
    const { getByText } = render(<Todo deleteTodo={deleteMock} />);
    const deleteButton = getByText("X");
    fireEvent.click(deleteButton);
    // const updateButton = getByText("Update!");
    // fireEvent.click(updateButton);
    expect(deleteMock).toHaveBeenCalled();
});
