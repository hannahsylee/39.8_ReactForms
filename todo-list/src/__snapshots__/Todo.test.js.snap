// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <li />
    <button>
      X
    </button>
    <button>
      Edit Task
    </button>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot when editing 1`] = `
<DocumentFragment>
  <div>
    <form>
      <input
        id="task"
        name="task"
        type="text"
        value=""
      />
      <button>
        Update!
      </button>
    </form>
  </div>
</DocumentFragment>
`;
